include/dnscpp/udp.h

```diff
@@ -60,70 +60,140 @@ class Udp : private Monitor
-     */
-    int setintopt(int optname, int32_t optval);
     
 private:
     /**
      *  event loop
      *  @var Loop*
      */
     Loop *_loop;
-    
+
     /**
-     *  The filedescriptor of the socket
-     *  @var int
+     *  The object that is interested in handling responses
+     *  @var Handler*
      */
-    int _fd = -1;
-    
+    Handler *_handler;
+
     /**
-     *  User space identifier of this monitor
-     *  @var void *
+     *  Private helper struct to represent a socket
      */
-    void *_identifier = nullptr;
+    struct Socket final : private Monitor
+    {
+        /**
+         *  Pointer to a Udp object
+         *  @var Udp*
+         */
+        Udp *parent = nullptr;
+
+        /**
+         *  User space identifier of this monitor
+         *  @var void *
+         */
+        void *identifier = nullptr;
+
+        /**
+         *  The filedescriptor of the socket
+         *  @var int
+         */
+        int fd = -1;
+
+        /**
+         *  Constructor does nothing but store a pointer to a Udp object.
+         *  Sockets are opened lazily
+         *  @param parent
+         */
+        Socket(Udp *parent);
+
+        /**
+         *  Closes the file descriptor
+         */
+        ~Socket();
+
+        /**
+         *  Check whether this socket has a valid file descriptor
+         *  @return bool
+         */
+        bool valid() const noexcept { return fd > 0; }
+
+        /**
+         *  Implements the Monitor interface
+         */
+        void notify() override;
+
+        /**
+         *  Send a query over this socket
+         *  @param  ip IP address to send to. The port is always assumed to be 53.
+         *  @param  query  The query
+         *  @return bool
+         */
+        bool send(const Ip &ip, const Query &query);
+
+        /**
+         *  Send a query to a certain nameserver
+         *  @param  address     target address
+         *  @param  size        size of the address
+         *  @param  query       query to send
+         *  @return bool
+         */
+        bool send(const struct sockaddr *address, size_t size, const Query &query);
+
+        /**
+         *  Open the socket
+         *  @param  version     IPv4 or IPv6. You must ensure this stays consistent over multiple requests (@todo)
+         *  @param  buffersize  The buffersize
+         *  @return bool
+         */
+        bool open(int version, int buffersize);
+
+        /**
+         *  Close the socket
+         *  @return bool
+         */
+        bool close();
+
+        /**
+         *  Helper method to set an integer socket option
+         *  @param  optname
+         *  @param  optval
+         */
+        int setintopt(int optname, int32_t optval);
+    };
 
     /**
-     *  The object that is interested in handling responses
-     *  @var Handler*
+     *  Socket needs access to the loop and the buffersize
      */
-    Handler *_handler;
-    
+    friend class Socket;
```

Comment:
This is a code smell that the architecture / separation of concerns is wrong. Do we need a helper class in the first place?

Reply:
> Do we need a helper class in the first place?

Yes. We need to keep track of the file descriptor, as well as the identifier for the loop. It's nice to have this pack of data together.

> This is a code smell that the architecture / separation of concerns is wrong.

I don't have a problem with private structs being friends with encompassing classes. But I suppose we can pass the loop and buffersize to it instead.